### PR TITLE
microverse rails: authentication

### PR DESCRIPTION
**Pull request by mahmoud717**
in this project I created a rails app that does the following:
1- added authentication for users.
2- when a user is logged in he can create a post and see the other posts with the poster name.
3- if the user is not logged in he can only see the posts without any names.

there are some issues with the commit history that I had to neglect because I created the repo from a template that had to branches( master & development ) and worked on the development branch directly, because of that I can't merge them together because they have no common commit history.

i will take care of that the next time..

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,29 +1,25 @@
...
   # GET /posts/1
   # GET /posts/1.json
-  def show
-  end
+  def show; end
 
   # GET /posts/new
-  def new
-    @post = current_user.posts.build
+  def new
```
Comment (voscarmv) [inline app/controllers/posts_controller.rb]: Create a new `Post` object here with `@post = Post.new`.
[Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_new_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication) :
image

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work! 🖥️ 👀 ✨ 
## However, some changes are still required... 🧰 🔨 🔧 

- Please add a sign-up button to the page layout to make it easier for users to sign-up when beginning to use the website.
- Please read my code comments for more information on fixing your code.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (mahmoud717) [reply app/controllers/posts_controller.rb]: I get your point, put, posts are created on my index page and use the index action, so I have to create a new post instance every time I open the page, so there is no use of having a new action at all.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team    
#### Great Job so far, everything works fine, but there are a few more things required. 
### General 
## Post
###### `new`

* [ ] it initializes an empty post `@post = Post.new` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_new_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by mahmoud717** [comment]
Guys the odin project it self said that if u don't wanna follow the exact steps then u don't have to.
U comment that I don't have a new action where I don't have a new view in the first place.
I'll say it again, the posts are created in the index page.
let me say one more time, the posts are created in the index page, that's the way I setup the project, and that's why I put the `@post = current_user.posts.build` in the index action.

**Review by MarioDena — CHANGES_REQUESTED** [review]
Hello! 

## Changes required 🔴 

I get your point totally, however your project can't get approved unless it containst every mandatory requirement implemented. You are indeed following the Odin's project guidelines, sadly this requirement is mandatory on Microverse not Odin's.

Think of this as a client asking you to do the Odin's project but with personalized milestones. 😄 

Please implement the requested changes and then resubmit for another review.

Cheers!  🎖️

**Comment by mahmoud717** [comment]
changes are implemented..

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -1,30 +1,18 @@
...
 
   # GET /posts
   # GET /posts.json
   def index
     @posts = Post.all
-    if user_signed_in?
-    @post = current_user.posts.build
-    end
+    @post = current_user.posts.build if user_signed_in?
     @users = User.all
```
Comment (maelfosso) [inline app/controllers/posts_controller.rb]: Sorry but `@users` is never used in the views, kindly remove it.

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for making the suggested changes.
However, there are still some issues.

1. Remove logic from views
https://github.com/mahmoud717/members-only/blob/05ad85150b1fbd58a2b2b22381464cca26f7f32b/app/views/posts/index.html.erb#L8-L12
https://github.com/mahmoud717/members-only/blob/05ad85150b1fbd58a2b2b22381464cca26f7f32b/app/views/posts/index.html.erb#L14-L18
https://github.com/mahmoud717/members-only/blob/05ad85150b1fbd58a2b2b22381464cca26f7f32b/app/views/layouts/application.html.erb#L14-L19
Move hose logic into view helpers.

2. Read also the comment left below

Kindly make the suggested changes and once done submit a new code review request
Thanks

Mael FOSSO

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (voscarmv) [reply app/controllers/posts_controller.rb]: @mahmoud717, sorry for not replying earlier.

The point here is not just cosmetic, it's a matter of good practice. 

A thorough explanation of why this is done in this way can be found here: https://stackoverflow.com/questions/5249151/what-is-the-purpose-of-the-new-controller-action-in-rails

It's a matter of maintainability of the code.

Hope this helps.

Regards!

**Comment by sumancrest0001** [comment]
**Status: Approved :100: :1st_place_medal: 
Congratulations @mahmoud717,** It is approved as you made all the changes requested by the previous TSE. Now, You can move to the next project.
[Optional] Before moving to the next project, you can remove all the json code because it is unnecessary for this project as you are not using them. 
**Happy coding !!!**

**Pull request merged**